// moonwave/moonwave.js
var createAtom = require('tiny-atom')
var createRouter = require('space-router')

module.exports = { moonwave: moonwave }

function moonwave (options) {
  if ( options === void 0 ) options = {};

  options.atom = options.atom || {}
  options.router = options.router || {}

  var app = {
    options: options,

    state: function (initialState) {
      options.atom.initialState = initialState
      return app
    },

    routes: function (routes, onTransition) {
      options.router.routes = routes
      options.router.onTransition = onTransition
      return app
    },

    evolve: function (evolve) {
      options.atom.evolve = evolve
      return app
    },

    actions: function (actions) {
      options.atom.actions = actions
      return app
    },

    render: function (fn) {
      options.render = fn
      return app
    },

    unrender: function (fn) {
      options.unrender = fn
      return app
    },

    mount: function (root) {
      options.root = root
      run(app)
      return app
    },

    unmount: function () {
      stop(app)
      return app
    }
  }

  return app
}

var run = function (app) {
  var options = app.options
  options.root = options.root || document.body
  options.atom.actions = options.atom.actions || {}
  options.atom.evolve = options.atom.evolve || defaultEvolve(options.atom.actions)
  validate(options)
  var router = app.router = createRouter(options.router.routes, options.router)
  var evolve = withRouting(options.atom.evolve, options.atom.actions, router)
  var onChange = debounce(app.options.render(app))
  var atom = app.atom = createAtom(options.atom.initialState, evolve, onChange, options.atom)
  var onTransition = options.onTransition || (function (route) { return atom.split({ route: route }); })
  router.start(onTransition)
}

var stop = function (app) {
  app.router.stop()
  app.options.unrender(app)
}

var withRouting = function (evolve, actions, router) { return function (get, split, action) {
  if (action.type === 'navigate') {
    router.push(action.payload.path, action.payload)
  } else {
    evolve(get, split, action, actions)
  }
}; }

var defaultEvolve = function (actions) {
  return function (get, split, action) {
    actions[action.type](get, split, action.payload)
  }
}

var debounce = function (fn) {
  var timeout
  return function () {
    var context = this
    var args = arguments
    var later = function () {
      timeout = null
      fn.apply(context, args)
    }
    clearTimeout(timeout)
    timeout = setTimeout(later, Math.round(1000 / 60))
  }
}

var validate = function (options) {
  if (!options.render) {
    throw new Error('Missing render implementation, provide with app.render((app) => (atom) => {})')
  }
  if (!options.unrender) {
    throw new Error('Missing unrender implementation, provide with app.unrender((app) => {})')
  }
  if (!options.router.routes) {
    throw new Error('Missing route map, provide with app.routes([[\'*\', App]])')
  }
}
